In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import seaborn as sns


In [9]:
df=pd.read_csv("/content/adult.csv",names=['age',"workclass","fnlwgt","education","education-num","marital-status","occupation","relationship","race","sex","capital-gain","capital-loss","hours-per-week","native-country","class"])

In [10]:
df.head(2)

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   

        marital-status        occupation    relationship    race    sex  \
0        Never-married      Adm-clerical   Not-in-family   White   Male   
1   Married-civ-spouse   Exec-managerial         Husband   White   Male   

   capital-gain  capital-loss  hours-per-week  native-country   class  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K

In [11]:
df.tail(2)

age      workclass  fnlwgt education  education-num  \
32559   22        Private  201490   HS-grad              9   
32560   52   Self-emp-inc  287927   HS-grad              9   

            marital-status        occupation relationship    race      sex  \
32559        Never-married      Adm-clerical    Own-child   White     Male   
32560   Married-civ-spouse   Exec-managerial         Wife   White   Female   

       capital-gain  capital-loss  hours-per-week  native-country   class  
32559             0             0              20   United-States   <=50K  
32560         15024             0              40   United-States    >50K

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  class           32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [14]:
df.describe()

age        fnlwgt  education-num  capital-gain  capital-loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours-per-week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000

In [16]:
df.shape

(32561, 15)

To find the null values

In [17]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
class             0
dtype: int64

#Conclusion
1. The size of data is 32561 rows and 15 column
2.Their are no null value present in any of the columns.
3.Their are 6 numerical data and 9 non numerical

In [19]:
numerical_column=list()
categorical_column=list()

for i in df.columns:
  if(df[i].dtype=='object'):
    categorical_column.append(i)
  else:
    numerical_column.append(i)

In [21]:
#mean of every numerical column
for i in numerical_column:
  print(round(np.mean(df[i]),2))

38.58
189778.37
10.08
1077.65
87.3
40.44


#Duplicate rows

In [23]:
df[df.duplicated()]

age          workclass  fnlwgt      education  education-num  \
4881    25            Private  308144      Bachelors             13   
5104    90            Private   52386   Some-college             10   
9171    21            Private  250051   Some-college             10   
11631   20            Private  107658   Some-college             10   
13084   25            Private  195994        1st-4th              2   
15059   21            Private  243368      Preschool              1   
17040   46            Private  173243        HS-grad              9   
18555   30            Private  144593        HS-grad              9   
18698   19            Private   97261        HS-grad              9   
21318   19            Private  138153   Some-college             10   
21490   19            Private  146679   Some-college             10   
21875   49            Private   31267        7th-8th              4   
22300   25            Private  195994        1st-4th              2   
22367   44            Private  367749      Bachelors             13   
22494   49   Self-emp-not-inc   43479   Some-college             10   
25872   23            Private  240137        5th-6th              3   
26313   28            Private  274679        Masters             14   
28230   27            Private  255582        HS-grad              9   
28522   42            Private  204235   Some-college             10   
28846   39            Private   30916        HS-grad              9   
29157   38            Private  207202        HS-grad              9   
30845   46            Private  133616   Some-college             10   
31993   19            Private  251579   Some-college             10   
32404   35            Private  379959        HS-grad              9   

            marital-status          occupation    relationship  \
4881         Never-married        Craft-repair   Not-in-family   
5104         Never-married       Other-service   Not-in-family   
9171         Never-married      Prof-specialty       Own-child   
11631        Never-married        Tech-support   Not-in-family   
13084        Never-married     Priv-house-serv   Not-in-family   
15059        Never-married     Farming-fishing   Not-in-family   
17040   Married-civ-spouse        Craft-repair         Husband   
18555        Never-married       Other-service   Not-in-family   
18698        Never-married     Farming-fishing   Not-in-family   
21318        Never-married        Adm-clerical       Own-child   
21490        Never-married     Exec-managerial       Own-child   
21875   Married-civ-spouse        Craft-repair         Husband   
22300        Never-married     Priv-house-serv   Not-in-family   
22367        Never-married      Prof-specialty   Not-in-family   
22494   Married-civ-spouse        Craft-repair         Husband   
25872        Never-married   Handlers-cleaners   Not-in-family   
26313        Never-married      Prof-specialty   Not-in-family   
28230        Never-married   Machine-op-inspct   Not-in-family   
28522   Married-civ-spouse      Prof-specialty         Husband   
28846   Married-civ-spouse        Craft-repair         Husband   
29157   Married-civ-spouse   Machine-op-inspct         Husband   
30845             Divorced        Adm-clerical       Unmarried   
31993        Never-married       Other-service       Own-child   
32404             Divorced       Other-service   Not-in-family   

                      race      sex  capital-gain  capital-loss  \
4881                 White     Male             0             0   
5104    Asian-Pac-Islander     Male             0             0   
9171                 White   Female             0             0   
11631                White   Female             0             0   
13084                White   Female             0             0   
15059                White     Male             0             0   
17040                White     Male             0             0   
18555                Black     Male         

#Count of duplicate rows

In [24]:
len(df[df.duplicated()])

24

#Avg age of people involved in datset

In [26]:
round(np.mean(df['age']))

39

#People belonging to which workclass

In [30]:
df['workclass'].value_counts()

 Private             22696
 Self-emp-not-inc     2541
 Local-gov            2093
 ?                    1836
 State-gov            1298
 Self-emp-inc         1116
 Federal-gov           960
 Without-pay            14
 Never-worked            7
Name: workclass, dtype: int64

#Education qualication of people

In [32]:
df['education'].value_counts()

 HS-grad         10501
 Some-college     7291
 Bachelors        5355
 Masters          1723
 Assoc-voc        1382
 11th             1175
 Assoc-acdm       1067
 10th              933
 7th-8th           646
 Prof-school       576
 9th               514
 12th              433
 Doctorate         413
 5th-6th           333
 1st-4th           168
 Preschool          51
Name: education, dtype: int64

#Avg number of hours working in a week


In [35]:
round(np.mean(df['hours-per-week']))

40

#People belonging to which country?

In [38]:
df['native-country'].value_counts()

 United-States                 29170
 Mexico                          643
 ?                               583
 Philippines                     198
 Germany                         137
 Canada                          121
 Puerto-Rico                     114
 El-Salvador                     106
 India                           100
 Cuba                             95
 England                          90
 Jamaica                          81
 South                            80
 China                            75
 Italy                            73
 Dominican-Republic               70
 Vietnam                          67
 Guatemala                        64
 Japan                            62
 Poland                           60
 Columbia                         59
 Taiwan                           51
 Haiti                            44
 Iran                             43
 Portugal                         37
 Nicaragua                        34
 Peru                             31
 

#Changing value of '<=50K' to 'A' and '>50K' to 'B' to make analysis easier using map function


In [42]:
df['class']=df['class'].map({" <=50K":"A"," >50K":"B"})

#Number of male and female in dataset

In [44]:
df['sex'].value_counts()

 Male      21790
 Female    10771
Name: sex, dtype: int64

#Does age plays an important role in deciding the class of people?

In [64]:
df.groupby('class').mean()['age'].reset_index()

class        age
0     A  36.783738
1     B  44.249841

In [77]:
#graphical reprsentation 
clas=df.groupby('class').mean()['age'].reset_index()['class']
age=df.groupby('class').mean()['age'].reset_index()['age']
px.bar(x=clas,y=age,labels={'x': 'class', 'y':'Average Age'})



#Handle '?' in column WORKING CLASS

In [81]:
df['workclass'].unique()

array([' State-gov', ' Self-emp-not-inc', ' Private', ' Federal-gov',
       ' Local-gov', ' ?', ' Self-emp-inc', ' Without-pay',
       ' Never-worked'], dtype=object)

In [85]:
df_class=df[df['workclass']==' ?']
len(df_class)

1836

Since 1800+ rows have '?' in them it is not feasible to drop all 1800 columns , instead we will be replacing it will 'Unkonwn' value

In [89]:
def replace_question_mark(a):
  if(a==' ?'):
    return 'unknown'
  return a

In [93]:
df['workclass']=df['workclass'].apply(replace_question_mark)

#Which working class people mostly belong to class A

In [96]:
df.groupby('workclass').count()['class'].reset_index()

workclass  class
0        Federal-gov    960
1          Local-gov   2093
2       Never-worked      7
3            Private  22696
4       Self-emp-inc   1116
5   Self-emp-not-inc   2541
6          State-gov   1298
7        Without-pay     14
8            unknown   1836

In [100]:
clas=df.groupby('workclass').count()['class'].reset_index()['class']
workclass=df.groupby('workclass').count()['class'].reset_index()['workclass']
px.bar(x=workclass,y=clas,labels={'x':"workclass",'y':'count of people'})

#which workclass people are the most hard working (max number of working hours)

In [105]:
df.groupby('workclass').mean()['hours-per-week'].reset_index()

workclass  hours-per-week
0        Federal-gov       41.379167
1          Local-gov       40.982800
2       Never-worked       28.428571
3            Private       40.267096
4       Self-emp-inc       48.818100
5   Self-emp-not-inc       44.421881
6          State-gov       39.031587
7        Without-pay       32.714286
8            unknown       31.919390

In [109]:
workclass=df.groupby('workclass').mean()['hours-per-week'].reset_index()['workclass']
work=df.groupby('workclass').mean()['hours-per-week'].reset_index()['hours-per-week']

px.histogram(x=workclass,y=work,labels={'x':"workclass",'y':"avg number of hours"})

#Does capital gain of people depend on race of people

In [114]:
df.groupby('race').mean()['capital-gain'].reset_index()

race  capital-gain
0   Amer-Indian-Eskimo    625.266881
1   Asian-Pac-Islander   1478.358037
2                Black    609.940461
3                Other    934.660517
4                White   1121.660375

In [115]:
px.scatter(x=df.groupby('race').mean()['capital-gain'].reset_index()['race'],y=df.groupby('race').mean()['capital-gain'].reset_index()['capital-gain'])

#Handling '?' value in   native country

In [125]:
def replace_mark(a):
  if(a==' ?'):
    return np.nan
  return a

In [130]:
df['native-country']=df['native-country'].apply(replace_mark)

#people living in which country and which working class sufers highest capital loss

In [171]:
df.groupby(['native-country','workclass']).sum().reset_index()[['native-country','workclass','capital-loss']].sort_values(by='capital-loss')[::-1]


native-country          workclass  capital-loss
208   United-States            Private       1660797
210   United-States   Self-emp-not-inc        265285
206   United-States          Local-gov        215188
209   United-States       Self-emp-inc        147175
213   United-States            unknown         98993
..              ...                ...           ...
129            Laos            Private             0
128            Laos          Local-gov             0
127            Laos        Federal-gov             0
126           Japan            unknown             0
0          Cambodia        Federal-gov             0

[225 rows x 3 columns]

#What is the percentage of white people in the data

In [186]:
df['race'].value_counts(normalize=True)*100

 White                 85.427352
 Black                  9.594300
 Asian-Pac-Islander     3.190934
 Amer-Indian-Eskimo     0.955130
 Other                  0.832284
Name: race, dtype: float64

In [195]:
px.pie(df['race'],names='race')